text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,802 @@
+general knowledge of ancillary data and the Unix sockets API?
+
+## Examples
+
+Sending a file descriptor:
+
+```rust
+use std::os::unix::net::{AncillaryData, MessageSender, UnixStream};
+use std::os::fd::AsFd;
+
+fn send_file(stream: &UnixStream, file: std::fs::File) -> std::io::Result<()> {
+    let mut ancillary = AncillaryData::new();
+    ancillary.add_file_descriptors(&[
+        file.as_fd(),
+    ]);
+
+    MessageSender::new(stream, b"\x00")
+        .ancillary_data(&mut ancillary)
+        .send(&stream)?;
+    Ok(())
+}
+```
+
+Receiving a file descriptor:
+
+```rust
+use std::os::unix::net::{AncillaryData, MessageReceiver, UnixStream};
+
+fn recv_file(stream: &UnixStream) -> std::io::Result<std::fs::File> {
+    // TODO: expose CMSG_SPACE() in a user-friendly way.
+    const ANCILLARY_CAPACITY: usize = 100;
+
+    let mut ancillary = AncillaryData::with_capacity(ANCILLARY_CAPACITY);
+    let mut buf = [0u8; 1];
+    MessageReceiver::new(stream, &mut buf)
+        .ancillary_data(&mut ancillary)
+        .recv(&stream)?;
+
+    // TODO: error handling (std::io::Error if not enough FDs returned)
+    let mut owned_fds = ancillary.take_owned_fds().unwrap();
+    let sent_fd = owned_fds.swap_remove(0);
+    Ok(sent_fd.into())
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Control messages
+
+Each control message is represented as a `ControlMessage` struct, which holds
+a reference to the underlying data.
+
+```rust
+struct ControlMessage<'a>;
+
+impl ControlMessage<'a> {
+    fn new(
+        cmsg_level: c_int,
+        cmsg_type: c_int,
+        data: &'a [u8],
+    ) -> ControlMessage<'a>;
+
+    // Opaque platform-specific integers; same as `struct cmsghdr` fields.
+    fn cmsg_level(&self) -> c_int;
+    fn cmsg_type(&self) -> c_int;
+
+    // The type-specific data of this control message.
+    fn data(&self) -> &[u8];
+
+    // Whether this control message is truncated, such as by being received
+    // into a too-short buffer.
+    fn truncated(&self) -> bool;
+}
+```
+
+A `&ControlMessages` represents a buffer full of control messages. It can be
+inspected and iterated over to obtain `ControlMessage` values.
+
+
+```rust
+struct ControlMessages;
+
+impl ControlMessages {
+    fn as_bytes(&self) -> &[u8];
+    fn is_empty(&self) -> bool;
+    fn len(&self) -> usize;
+}
+
+impl<'a> IntoIterator for &'a ControlMessages {
+    type Item = ControlMessage<'a>;
+    type IntoIter = ControlMessagesIter<'a>;
+}
+
+struct ControlMessagesIter<'a> { ... }
+
+impl<'a> Iterator for ControlMessagesIter<'a> {
+    type Item = ControlMessage<'a>;
+}
+```
+
+A `ControlMessagesBuf` is the owned variant of `ControlMessages`. It exposes a
+subset of the `Vec` capacity management functions, with a public API that only
+allows operations that won't reduce its length (since that would risk losing
+information about received file descriptors).
+
+```rust
+struct ControlMessagesBuf;
```

Comment:
The `ControlMessagesBuf` capacity is in bytes, rather than messages -- and some message types are dynamically sized, for example `SCM_RIGHTS` is semantically a `(usize, [i32])` tuple.

I think it would be possible to have an `AncillaryData` backed by a `&mut [u8]`, but when writing the draft implementation I found it difficult to use. If the number of heap allocations is a concern, reusing an `AncillaryData` between socket calls is probably more practical than stack allocation.